### PR TITLE
Rethinking minimal example and defaults (RFC)

### PR DESCRIPTION
**Pull request by joallard**
I always appreciate when a library takes time to consider sensible defaults without needing configuration, allowing me to start quickly.

I would suggest, first off, to document the most minimal case that this plugin can be used. This is what this PR attempts. We could then tack on changing defaults (and schedule for minor/major revisions).

It would:
* put default essential options in comment
* include "required" (or strongly encouraged) options in the example

Some thoughts about the defaults:
* A sensible default for `el` would be a nice-to-have. This could be `.pickr`, `.js-pickr`, etc. Although specifying the element is not a big hassle. (Currently, no `el` throws an error)
* Documenting the default for `theme`, currently `classic`. See note below.
* Indicating a sensible path for `save` and `comparison`. See #195.
* Indicating a component needs to be set or choosing defaults. In my opinion, hue and preview are good picks.

About the theme: I wonder whether it's feasible to automatically detect the theme. eg using `getComputedStyle` on ready, and setting `.dummy-selector{ z-index: 1 }` (where z-index is the theme ID)

What do you think?

### CODE REVIEW

**Comment by joallard** [comment]
Some other surprises about the color callback:
* I expected `on('save', (color) => ...)` to be a browser-usable color. Maybe rename that parameter to `hsvaColorObj` or something similar
* I would also suggest aliasing `hsva.toString()` to `hsva.toHEXA().toString()` (currently it outputs `[object Object]`.

**Comment by Simonwep** [comment]
_Whups, accidentally clicked "ready for review"..._

First of all, thank you for spending time thinking about how the docs could be improved, really appreciate it, especially as I'm currently kinda occupied and cannot spend that much time working on it :)

#### Documentation
I absolutely agree to all your points, I think we're on a point where it's rather difficult to start since there are, compared to the first releases, many options!

Defining a default selector for `el` might be cool or confusing, not sure about that one though.

#### Theme detection
Regarding automatically detecting the theme I'd rather let the user configure it via js. Operations where style-recalculations are required are always kinda heavy to handle (performance) and currently, the project-core itself was written in my first year with js, the performance is already not that good (If you look at the boot-up time on mobile devices or with 4x CPU throttling).

#### `hsva` color object
About the color-callback. The thing is it is basically just a color which _can_ be expressed in many ways such as HEX, RGBA and so on. So I'd rather keep it as-is. `hsva.toString()` - What kind of return-value would you expect? 

Considering that it's possible to choose the format via `hsva.toHEXA().toString()` this could be a short way for quick usage and therefore return it in `rgba` format (To maximize compatibility, including the alpha-channel) - actually a pretty nifty idea :)

> And I'm totally open for any kind of refactoring - as I said this project wasn't made with the thought that I'd be that big at some point, hence there's quite a lot of room for improvement (basically the entire code-base).

**Comment by joallard** [comment]
Points about `el` and theme detection make sense, I don't have strong opinions about those.

The only thing I could point out about the `el` default is that it would be the most non-breaking of changes. At the moment, *everyone* has to specify `el`. The default would only apply for people not providing `el`. That's the beauty of convention over configuration, but remaining configurable ;)

A reason I like `.js-pickr` is pure opinion, granted, but comes after reading [Decouple JavaScript classes from CSS ones by using prefix](https://coderwall.com/p/qktuzw/decouple-javascript-classes-from-css-ones-by-using-prefix) arguing for `.js-` prefixes in classes:

> By using js- prefix for classes in DOM elements you decouple functional aspects of these elements from CSS ones. An example:
> 
> `<a class="btn js-popup">Click me!</a>`
> 
> Now:
> 
> *   as css coder, you can easily change anchor design by changing btn class and not worry it'll break pop up behavior.
> *   and conversely, as JavaScript developer you can remove popup behavior and never break any existing CSS rules.
>

I'll see whether that one sticks in my book ;) 

About the color object, I think one of the main use cases is to pass the returned value to set a CSS property. (eg. `element.style.backgroundColor = hsva.foo().bar()`)

I find what the user expects is often what the API should strive to be (principle of least surprise).

* I had first tried: `(color) => element.style.backgroundColor = color`
* Then debugged and saw it was an object, tried: `(hsva) => element.style.backgroundColor = hsva.toString()`
* Then went to see the docs and got to `toFormat.toString`

I'm seeing we have an option `defaultRepresentation`:

* Maybe calling `hsva.toString()` would be the equivalent of calling `hsva.to<DefaultRepresentation>().toString()`? 
* Or maybe `hsva.toString()` should default to a consistent format (HEX is classic, RGB reflects pixels, HSV is a good mental model)?

Maybe we could add aliases `hsva.formatHsl()`, like (d3-color)[https://github.com/d3/d3-color] does and document that? I think the d3-color API could be a good inspiration!

*By the way, my love of D3 and Observable led me here, the [main color picker input](https://observablehq.com/@shaunlebron/color-picker) in Observable uses Pickr*

About the other defaults, let me suggest this, counterproposals welcome:

* If `save: false` (default), and no swatches are there, `comparison` defaults to `false`
* If `comparison: true`, `save` defaults to `true`

(This could be scheduled for the next major release only)

* If `components` is empty, it defaults to `{hue: true, preview: true}` (We don't expect components to be empty; minor change)

I'm happy to integrate the result of our discussions in the new Readme

**Comment by Simonwep** [comment]
#### Regarding `el`
I like the idea of prefixing elements with js-related usage, I think we could do that - you decide as It's very seldom that I work with vanilla-js whereby I'm more or less inexperienced of these coding-conventions :) 

#### The `color` object
`(color) => element.style.backgroundColor = color` looks way better, default representation could lead to `CMYK` or `HSV` strings which are not supported by browsers, let's stick to `HSL` which  is [widely supported](https://caniuse.com/#feat=mdn-css_types_color_hsl)?!

This could be easily done by implementing `toString` [here](https://github.com/Simonwep/pickr/blob/master/src/js/utils/hsvacolor.js#L12):
```js
toString() {
    return that.toHSLA().toString();
}
```

#### Aliases
I'd actually stick to the current way, maybe this could be refactored but it should be kept in a concise and consistent style like `to[name].toString()`. Maybe we could add a `.format()` function to each array returned by one of the color functions. This way it'd look like `toHEXA().format()`.
Adding for each `format` function separate methods would increase the bundle size noticeable.

#### Defaults...
I agree to all three proposals, they all make perfect sense and would simplify things a lot. The only thing which might be an obstacle to it could be the amount of documentation to be added what again does not always simplify the usage :/

I just took a look at Observable - `pickr-widget@0.2.0`; now **that's** a way outdated and not even stable version 😅

> I'm happy to integrate the result of our discussions in the new Readme

Of course, thank you very much 🤗

**Comment by joallard** [comment]
In searching for what was the ideal solution for the default 'el' param, I came to the realization that using `<color-picker>` is probably the most sensible suggestion.

* It is minimal (opens the door to not using any js at all)
* It is semantic
* It is implementation agnostic. Include one color-picker library and off you go.

This is still pretty experimental, so commits can be branched off at will. But it would open the door to using new techs such as the shadow DOM (encapsulation of CSS) and so on.

In an ideal world, low-configuration would be just dumping the element in the HTML, adding configuration attributes if needed; if more customization is needed, just explicitly make the call to `.create()`.

In the meantime, just proposing it as a default is pretty cool imo.

---

So. Just by tweaking the docs and defaults a bit, we achieve a lot. 

There is still the save/comparison situation which I'm not totally grokking at the moment. It seems to me like they're completely contradictory and should be just one option. I could use some guidance on this.

There is also the deep merge concern on the 'components' option. I didn't want to implement a deep merge nor bring in a new library, so I just added some warnings. 20% of efforts for 80% of results.

Feedback welcome on all of this!

File: README.md [hunk 1]
```diff
@@ -104,6 +104,11 @@ const pickr = Pickr.create({
         'rgba(255, 193, 7, 1)'
     ],
 
+    // either
+    save: true,
```
Comment (Simonwep) [inline README.md]: Do you mean `components.interaction.save`? Because there is no root `save` option (I think)...

**Comment by Simonwep** [comment]
I totally agree to all points mentioned, we should definitely keep the `components`-default and readme-updates / clarifications.

About `<color-picker>` - it looks like a _real_ web-component and using a custom-element just for selection purposes is kinda strange. As you said, the entire DOM-Staff could be put into a open-shadow-dom (so people can modify its styles) and pickr itself could be "refactored" as web-component (As of `v2.0.0` - this would be really cool :D)

The save/comparison thingy is really just to provide maximal configuration possibilities. I don't know if it actually makes sense but hey - maybe someone will make use of it :)

Wrapping pickr into a web-component shouldn't be that difficult / time-consuming and (after my upcoming finals) I think I could do that - I really appreciate your efforts so far, but let us - in favor of web-components, not blur the line between the "old way" to initialize it (via `el`) and the new way by just doing something like

```html
<pickr color="#fff" theme="classic">
```

> Maybe we could, to clean up things a bit by merging 718544d89502f4067d3b875940806fd5a9653728 
and 15d8cee15ae0b60a9a48088ee98a0275f9d7d3be (is that possible?).

What do you think about this idea? If further configuration is needed we could still add a manual mount-method or add something like `config="<JSON-Config>"` 🤔 (I'm not sure what's the best way in this case)

**Comment by Simonwep** [comment]
I'll close this because this project is now more or less [feature-locked](https://github.com/Simonwep/pickr#status-of-this-project). Maybe I'll do a complete rewrite at some point but for now I won't have the time to do any major changes anymore.  Thank you @joallard for your ideas, if I every do this again I'll come back to what we've discussed here :)

**Comment by joallard** [comment]
Yep, I think I got a bit hungry on the scope creep with this PR. If I get the chance, I'll try to come back with an easy-to-merge PR about the docs. Thanks for the clarifications!

**Pull request closed without merging**